dialogue-core/src/main/java/com/palantir/dialogue/core/RetryOtherValidatingChannel.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.net.HttpHeaders;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.RateLimiter;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import com.palantir.dialogue.futures.DialogueFutures;
+import com.palantir.logsafe.UnsafeArg;
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import javax.annotation.CheckForNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class RetryOtherValidatingChannel implements Channel {
+
+    private static final Logger log = LoggerFactory.getLogger(RetryOtherValidatingChannel.class);
+    private static final RateLimiter VALIDATION_FAILED_LOGGING_LIMITER = RateLimiter.create(1);
+
+    private final Channel delegate;
+    private final List<String> hosts;
+    private final Consumer<String> failureReporter;
+
+    RetryOtherValidatingChannel(Channel delegate, List<String> hosts) {
+        this(delegate, hosts, RetryOtherValidatingChannel.failureReporter(hosts));
+    }
+
+    @VisibleForTesting
+    RetryOtherValidatingChannel(Channel delegate, List<String> hosts, Consumer<String> failureReporter) {
+        this.delegate = delegate;
+        this.hosts =
+                hosts.stream().map(RetryOtherValidatingChannel::strictParseHost).collect(Collectors.toList());
```

Comment:
I feel like we're missing validating that somewhere earlier, this validation is done in a bunch of places now. But probably not for this PR

Reply:
java URL and URI types aren't very good about keeping up with standards (with good reason I imagine, since changes are likely to break users) so we don't necessarily standardize on them.

Should we use a set to avoid unnecessary iteration? for-loop over strings is probably faster for small values, but we don't need to optimize the 308 path.
```suggestion
                hosts.stream().map(RetryOtherValidatingChannel::strictParseHost).collect(ImmutableSet.toImmutableSet());
```

Reply:
Done